src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -534,51 +535,62 @@ private async ValueTask<ProtocolToken> ReceiveBlobAsync<TIOAdapter>(TIOAdapter a
             }
 
             // At this point, we have at least one TLS frame.
-            if (_lastFrame.Header.Type == TlsContentType.Alert)
+            switch (_lastFrame.Header.Type)
             {
-                if (TlsFrameHelper.TryGetFrameInfo(_handshakeBuffer.ActiveReadOnlySpan, ref _lastFrame))
-                {
-                    if (NetEventSource.Log.IsEnabled() && _lastFrame.AlertDescription != TlsAlertDescription.CloseNotify) NetEventSource.Error(this, $"Received TLS alert {_lastFrame.AlertDescription}");
-                }
-            }
-            else if (_lastFrame.Header.Type == TlsContentType.Handshake)
-            {
-                if (_handshakeBuffer.ActiveReadOnlySpan[TlsFrameHelper.HeaderSize] == (byte)TlsHandshakeType.ClientHello &&
-                    (_sslAuthenticationOptions!.ServerCertSelectionDelegate != null ||
-                    _sslAuthenticationOptions!.ServerOptionDelegate != null))
-                {
-                    TlsFrameHelper.ProcessingOptions options = NetEventSource.Log.IsEnabled() ?
-                                                                TlsFrameHelper.ProcessingOptions.All :
-                                                                TlsFrameHelper.ProcessingOptions.ServerName;
-
-                    // Process SNI from Client Hello message
-                    if (!TlsFrameHelper.TryGetFrameInfo(_handshakeBuffer.ActiveReadOnlySpan, ref _lastFrame, options))
+                case TlsContentType.Alert:
+                    if (TlsFrameHelper.TryGetFrameInfo(_handshakeBuffer.ActiveReadOnlySpan, ref _lastFrame))
                     {
-                        if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"Failed to parse TLS hello.");
+                        if (NetEventSource.Log.IsEnabled() && _lastFrame.AlertDescription != TlsAlertDescription.CloseNotify) NetEventSource.Error(this, $"Received TLS alert {_lastFrame.AlertDescription}");
                     }
-
-                    if (_lastFrame.HandshakeType == TlsHandshakeType.ClientHello)
+                    break;
+                case TlsContentType.Handshake:
+                    // During renegotiation the client hello is encrypted. Possibly race condition when the encrypted client hello flag is 0x01 but the content of the message is encrypted.
```

Comment:
We know if we are doing renegotiation by the _isRenego flag, right? Should we be checking that here and skip this logic?

Reply:
In practice I see always 0 there but I could not find part of the spec that would deterministicly explain it. Since we do have the _isRenego I added it to the check and I removed the comment.